src/heretic/r_plane.c

```diff
@@ -507,12 +510,12 @@ void R_DrawPlanes(void)
             }
             continue;
         }
-
-        //
+        swirling = (flattranslation[pl->picnum] == -1); //[crispy] adapting swirling to heretic branch from doom's one
         // regular flat
-        //
-        lumpnum = firstflat + flattranslation[pl->picnum];
-
+        lumpnum = firstflat + (swirling ? pl->picnum : flattranslation[pl->picnum]);
+        // [crispy] add support for SMMU swirling flats
+        ds_source = swirling ? R_DistortedFlat(lumpnum) : W_CacheLumpNum(lumpnum, PU_STATIC);
+        if (!swirling){
```

Comment:
This is redundant. You might want to turn this into something like `if (!swirling) {..} else {ds_source = R_DistortedFlat(lumpnum);}`.